File: app/src/main/java/com/example/mrakopediareader/MrakopediaWebViewClient.java
package com.example.mrakopediareader;

import android.graphics.Bitmap;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import io.reactivex.rxjava3.subjects.PublishSubject;

class MrakopediaWebViewClient extends WebViewClient {
    private PublishSubject<Boolean> loadingSubject$ = PublishSubject.create();

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        super.onPageStarted(view, url, favicon);
        this.loadingSubject$.onNext(true);
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        super.onPageFinished(view, url);
        this.loadingSubject$.onNext(false);
    }

    public PublishSubject<Boolean> getLoadingSubject() {
        return this.loadingSubject$;
    }
}

File: app/src/main/java/com/example/mrakopediareader/viewpage/MrakopediaWebViewClient.kt
package com.example.mrakopediareader.viewpage

import android.graphics.Bitmap
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient

internal class MrakopediaWebViewClient(
    private val onStartLoading: () -> Unit,
    private val onFinishLoading: () -> Unit
) : WebViewClient() {
    override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
        super.onPageStarted(view, url, favicon)
        onStartLoading()
    }

    override fun onReceivedError(
        view: WebView?,
        request: WebResourceRequest?,
        error: WebResourceError?
    ) {
        super.onReceivedError(view, request, error)
        view?.apply {
            val htmlData = """
                <!doctype html>
                <html>
                  <body>
                    <div style="height: 100vh; display: flex; align-items: center; justify-content: center">
                      Невозможно загрузить страницу :(
                    </div>
                  </body>
                </html>
            """.trimIndent()
            loadUrl("about:blank")
            loadDataWithBaseURL(null, htmlData, "text/html", "UTF-8", null)
            invalidate()
        }
    }

    override fun onPageFinished(view: WebView, url: String) {
        super.onPageFinished(view, url)
        onFinishLoading()
    }
}